### PR TITLE
Add logging system

### LINKED ISSUE #4: Add logging system
### What?
We would like to add optional logging system, which will inform about all module/pipeline-related events. E.g. module's bootstrap/teardown or beginning request processing.

### Why?
To improve debugging and add easily integrable verbose mode.

### How?
By adding an option, which will enable/disable verbose mode. MAGDA will automatically print out logging messages while in verbose mode. The target (console/file/anything else) will be configurable

### PR DESCRIPTION
**Pull request by jamebs**
### Overview
Loggers are configured via `MagdaLogger.Config` class. You can configure the output stream (logging, print, any callable), format and filter records (e.g. skip *events* like starting processing). Loggers are disabled by default. Under the hood, they are chained and attached to a proper class metadata (like pipeline/group/module names, replica index etc.).

You can use these loggers also to print own records. The right logger is passed as a new argument `logger` with a type `MagdaLogger` to all overridden methods (`run`, `bootstrap` and `teardown`).

#### Todo
- [x] add configurable logging system:
  - [x] sequential pipelines
  - [x] parallel pipelines
  - [x] support replicated groups
- [x] Add option to disable colors
- [x] Re-format output (add option for join parts, e.g. `(g1 : 2 : m1)` = module `m1` from 2nd `g1` group replica)
- [x] add unit-tests

Resolves #4

### CODE REVIEW

File: examples/modules/a.py [hunk 1]
```diff
@@ -3,37 +3,34 @@
 
 from magda.module import Module
 from magda.decorators import register, accept, finalize, produce
```
Comment (DanielPopek) [inline examples/modules/a.py:5]: I know, those are only examples and they have already existed, but still, files named "a.py", "b.py", "c.py" sound ambiguous :) Please change them at least to "module_a.py",  "module_b.py" etc.

File: magda/config_reader.py [hunk 2]
```diff
@@ -1,13 +1,17 @@
-import yaml
+from __future__ import annotations
+
```
Comment (DanielPopek) [inline magda/config_reader.py:2]: Please note the conflicts with the current version ( exposing results in config file ).

File: magda/pipeline/parallel/parallel_pipeline.py [hunk 3]
```diff
@@ -31,14 +32,20 @@ def _list_group_dependencies(self, group: str) -> Set[str]:
             if m.group == group and dep.group != group
         ])
 
-    async def build(self, context=None, shared_parameters=None) -> Runtime:
+    async def build(
+        self,
+        context=None,
+        shared_parameters=None,
+        *,
+        logger: Optional[MagdaLogger.Config] = None,
```
Comment (DanielPopek) [inline magda/pipeline/parallel/parallel_pipeline.py:40]: Please change the arg name to `logger_config` - will be consistent with the naming in other classes, where config is passed to the constructor

File: magda/pipeline/graph.py [hunk 4]
```diff
@@ -98,16 +108,28 @@ async def run(
...
+        self._logger = logger
         for module in self._modules:
-            await module._on_bootstrap()
+            logger = self._logger.chain(
+                module=MagdaLogger.Parts.Module(
+                    name=module.name,
+                    kind=module.__class__.__name__,
+                ),
+            )
+            logger.event('BOOTSTRAP')
```
Comment (DanielPopek) [inline magda/pipeline/graph.py]: Hardcoded strings:) Both with "TEARDOWN". I don't know if we want to change them, but they could be declared upwards, at the top of the file.

File: test/utils/logger/test_pipeline_logging.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        assert mock.call_count == 6
+        mock.assert_any_call(PartialOutput(f'{pipe_name}.*{module1.name}.*BOOTSTRAP'))
+        mock.assert_any_call(PartialOutput(f'{pipe_name}.*{module2.name}.*BOOTSTRAP'))
+        mock.assert_any_call(PartialOutput(f'{pipe_name}.*{module1.name}.*RUN'))
+        mock.assert_any_call(PartialOutput(f'{pipe_name}.*{module2.name}.*RUN'))
+        mock.assert_any_call(PartialOutput(f'{pipe_name}.*{module1.name}.*{MESSAGE}'))
+        mock.assert_any_call(PartialOutput(f'{pipe_name}.*{module2.name}.*{MESSAGE}'))
+
+    async def test_should_log_pipeline_with_colors(self):
```
Comment (DanielPopek) [inline test/utils/logger/test_pipeline_logging.py:93]: As LoggerConfig can be passed via ConfigReader, could you also add one test checking if the config is set correctly while reading the pipeline from a yaml file :) ? I don't insist, however.

**Review by DanielPopek — COMMENTED** [review]
Tests, Lint, examples ran and logic checked. I think it is a very neat and colorful solution :)

File: examples/example2.py [hunk 6]
```diff
@@ -20,7 +19,7 @@ async def demo(cls):
         await example.run()
 
     async def build(self, prefix: str = '{CTX}'):
-        builder = ParallelPipeline()
+        builder = ParallelPipeline(name='Example')
```
Comment (lukaszsus) [inline examples/example2.py:23]: You have added a pipeline name here but not in other examples from `examples/` directory. It is not a huge problem, because it show that the `name` argument is not necessary but I am not sure that you did it on purpose. I would rather choose the same convention in all examples but the final decision is up to you. :-)

File: magda/module/module.py [hunk 7]
```diff
@@ -106,5 +106,5 @@ def build(self, context=None, shared_parameters=None):
             context=context,
             shared_parameters=shared_parameters,
             is_regular_module=self.is_regular_module,
-            parameters=self._parameters
+            parameters=self._parameters,
```
Comment (lukaszsus) [inline magda/module/module.py:113]: I think we don't want to have this comma in the end of this line.

File: magda/module/module.py [hunk 8]
```diff
@@ -95,7 +95,7 @@ def expose_result(self, name: Optional[str] = None) -> Module:
         self._exposed = name if name is not None else ''
         return self
 
-    def build(self, context=None, shared_parameters=None):
+    def build(self, *, context: Any = None, shared_parameters: Any = None):
```
Comment (lukaszsus) [inline magda/module/module.py:102]: This `*` can cause backward incompatibility. I know that probably we are the main user of magda but still this repository is public so I am not convinced that we want to introduce such a change without changing a major version of magda.

File: magda/module/runtime.py [hunk 9]
```diff
@@ -33,24 +33,46 @@ def __init__(
...
             self._context = self._context()
 
+        signature = inspect.signature(self.bootstrap)
+        parameters = [p.name for p in signature.parameters.values()]
+        fn_kwargs = {
+            key: value
+            for key, value in kwargs.items()
+            if key in parameters
+        }
+
```
Comment (lukaszsus) [inline magda/module/runtime.py]: I can see that you are filtering 'logger' out here if it is not expected by `bootstrap`. I appreciate the way that you implemented this but it is like a 'magic' happens here because looking only at this file I have no idea what's going on here. Moreover, having `bootstrap` and `_on_bootstrap` can be a little confusing for developers not familiar with `magda`. Maybe, a little docstring like 'A wrapper for a boostrap invoked by graph in runtime.' would be helpful. I know that every kind of comment is always a problem so I do not insist on this, I just think it could be helpful and simple and other `ModuleRuntime` methods are provided with docstrings. What do you think? :-)

File: magda/pipeline/graph.py [hunk 10]
```diff
@@ -63,16 +64,17 @@ def _get_module(self, module_name):
...
     def modules(self) -> List[Module.Runtime]:
         return self._modules.copy()
 
     async def run(
         self,
-        request: Any,
+        request,
         results: List[Module.Result] = [],
-        is_regular_runtime: bool = True,
+        is_regular_runtime=True,
```
Comment (lukaszsus) [inline magda/pipeline/graph.py]: Why did you remove typehints here?

File: magda/pipeline/graph.py [hunk 11]
```diff
@@ -135,25 +157,32 @@ async def _run_method_helper(
...
         is_regular_runtime: bool,
+        logger: MagdaLogger,
     ) -> Any:
         module_result = None
+        props = dict(
+            data=data,
+            request=request,
+            logger=logger,
+        )
+
```
Comment (lukaszsus) [inline magda/pipeline/graph.py]: We had a discussion with our workmates about initialization of python dictionaries: `dict()` vs `{}`. The conclusion of this discussion was that `{}` is more efficient. You can find some materials about that in the Internet, for example [here](https://deepsource.io/blog/python-performance-three-easy-tips/). Unfortunately, all Internet resource that I found consider an empty dictionary initialization so I have conducted my own experiment which tested initialization with some arguments. The result was 97 ns to 209 ns for `{}`. I think we do not have to fight for these nanoseconds but there was a chance that you do not know about this performance difference and I just wanted to share this knowledge with you. If you still prefer `dict()` to `{}` because you consider this syntax clearer, you can keep it. For me `{}` is more pythonic and since it is more efficient, I would rather choose it.

File: examples/example2.py (hunk 6 above)
Comment (jamebs) [reply examples/example2.py:23]: The name is optional, so I added it only once in the examples to show "how can you do that". I could add comment here, to explain that

File: magda/utils/logger/printers/uid.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class UidPrinter(BasePrinter):
+    def _with_colors(self, kind: str, name: str, group: str, replica: str) -> str:
+        return (
+            Fore.BLUE + f'{kind} ' + Style.BRIGHT
+            + f'({name}' + Fore.CYAN + group + Style.NORMAL + replica
+            + Fore.BLUE + Style.BRIGHT + ')'
+            + Fore.RESET + Style.NORMAL
+        )
```
Comment (lukaszsus) [inline magda/utils/logger/printers/uid.py:17]: I guess that this way of coloring text works in terminal but what about other types of loggers? How does it work with other loggers?

File: magda/module/module.py (hunk 7 above)
Comment (jamebs) [reply magda/module/module.py:113]: Yes - it's not necessary but it's recommended by most of the style-guides :smile: Unfortunately, python linters don't care about that, so I added commas [where missing] to have the same convention for the whole project

File: magda/module/module.py (hunk 8 above)
Comment (jamebs) [reply magda/module/module.py:102]: Yes, but:
1. MAGDA is still in pre-release version (`0.1`)
2. `Module.build` shouldn't be invoked from the outside the library - only `Pipeline` is allowed to run this method
3. #1 already introduced breaking change :smile:

File: magda/pipeline/graph.py (hunk 10 above)
Comment (jamebs) [reply magda/pipeline/graph.py]: My mistake. Anyway, the type is inferred correctly (due to setting the value to `True`), but I will revert that change

File: magda/pipeline/graph.py (hunk 11 above)
Comment (jamebs) [reply magda/pipeline/graph.py]: Good to know, thanks! :raised_hands:

File: magda/module/runtime.py (hunk 9 above)
Comment (jamebs) [reply magda/module/runtime.py]: This little "magic" keeps `bootstrap()` method backward-compatible with the existing modules implementation by filtering out **all** unknown (*new*) arguments - like `logger` as you noticed. You're right, it could be a little bit confusing, so docstring is a great idea!

File: magda/pipeline/parallel/parallel_pipeline.py (hunk 3 above)
Comment (jamebs) [reply magda/pipeline/parallel/parallel_pipeline.py:40]: The name `logger` is set here on purpose, because it's the highest level, thus User's interface. For the User, there is no difference between `logger`, `logger_config` and anything else, because User always uses only `MagdaLogger.Config`. So i decided on the shorter name. However, if it is a little bit confusing, I can change to *more verbose* name

File: test/utils/logger/printers/test_group.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        output = printer.flush(colors=False, group=None)
+        assert output is None or output == ''
+
+    def test_should_accept_extra_arguments(self):
+        printer = GroupPrinter()
+        group = 'GroupG1'
+        replica = 20
+
+        part = LoggerParts.Group(group, replica)
+        output = printer.flush(colors=False, group=part, extra=[1, 2, 3], msg='Test')
```
Comment (lukaszsus) [inline test/utils/logger/printers/test_group.py]: As far as I understand this code, `extra` and `msg` arguments should not impact on the logging message? Maybe, it would be a good idea to compare the output of `printer.flush(colors=False, group=part)` with the output of `printer.flush(colors=False, group=part, extra=[1, 2, 3], msg='Test')`?

File: test/utils/logger/printers/test_message.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        assert output.find(event) != -1
+        assert re.search(self.COLOR_REGEXP, output) is None
+
+    def test_should_print_event_with_colors(self):
+        printer = MessagePrinter()
+        event = 'TEST'
+        output = printer.flush(colors=True, msg=event, is_event=True)
+        assert output.find(event) != -1
+        assert re.search(self.COLOR_REGEXP, output) is not None
+
```
Comment (lukaszsus) [inline test/utils/logger/printers/test_message.py:50]: Would you like to test `test_should_accept_empty_event`? Not knowing the implementation details you cannot decide whether event can be empty or not. Now, you test only non-event message.

File: test/utils/logger/printers/test_group.py (hunk 13 above)
Comment (lukaszsus) [reply test/utils/logger/printers/test_group.py]: If you decide to implement this change, please also implement it to similar places in other tests.

File: test/utils/logger/printers/test_timestamp.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import re
+from magda.utils.logger.printers import TimestampPrinter
+
+
+class TestTimestampPrinter:
+    TIMESTAMP_REGEXP = r'(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2}).(\d{3})'
```
Comment (lukaszsus) [inline test/utils/logger/printers/test_timestamp.py]: You probably missed escaping for '.' here. I think it should be:
`r'(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})\.(\d{3})'`

Comment (lukaszsus) [reply test/utils/logger/printers/test_timestamp.py]: Does it really work correctly? I think that you cut off the last 3 characters [here](https://github.com/jamebs/magda/blob/feature/logging/magda/utils/logger/printers/timestamp.py#L12) :thinking:

**Review by lukaszsus — COMMENTED** [review]
Some additional minor comments.

File: magda/utils/logger/printers/uid.py (hunk 12 above)
Comment (jamebs) [reply magda/utils/logger/printers/uid.py:17]: These are standard ANSI escape codes so it should *just work*. However, there is an option in MAGDA Logger for disabling colors in case of any troubles.

File: test/utils/logger/printers/test_group.py (hunk 13 above)
Comment (jamebs) [reply test/utils/logger/printers/test_group.py]: Sounds great!

File: test/utils/logger/printers/test_timestamp.py (hunk 15 above)
Comment (jamebs) [reply test/utils/logger/printers/test_timestamp.py]: Good point with escaping - indeed, I didn't notice that.

About the 3 last characters: python microseconds (`%f`) generates 6 digits ([python docs](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes)), which is too precise for a logger I believe. So I cut last part (`[:-3]`) to leave only milliseconds

File: test/utils/logger/test_logger.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+    def test_format_should_be_overriden(self):
+        request = MagdaLogger.Parts.Request('RequestR1')
+
+        mock = MagicMock()
+        config = MagdaLogger.Config(output=mock, format=[])
+        logger = MagdaLogger.of(config).chain(request=request)
+
+        logger.info('Hello!')
+        mock.assert_called_once_with('')
```
Comment (lukaszsus) [inline test/utils/logger/test_logger.py]: It's not a big deal but you could be consistent with `test_loggers_should_be_chained`. I mean you could initialize `request_name` and `message` in separate lines or you can pass them to functions and methods without previous initialization but it would be better to do it the same way in both tests.

File: magda/pipeline/parallel/runtime.py [hunk 17]
```diff
@@ -46,6 +58,12 @@ def jobs(self) -> List[Job]:
     def closed(self) -> bool:
         return self._is_closed
 
+    async def _bootstrap(self):
```
Comment (lukaszsus) [inline magda/pipeline/parallel/runtime.py:61]: Sorry, I am a little confused. Why do you use this `_bootstrap` name instead of `bootstrap`?

**Review by lukaszsus — COMMENTED** [review]
First iteration of review finally finished. I had some small comments and questions. :-)

File: examples/example2.py (hunk 6 above)
Comment (lukaszsus) [reply examples/example2.py:23]: Ok :-) I understand your decision. :-)

File: magda/module/module.py (hunk 7 above)
Comment (lukaszsus) [reply magda/module/module.py:113]: Ok, I am surprised. PEP8 says that they are optional but I can understand the arguments for using it. :wink:

File: magda/module/module.py (hunk 8 above)
Comment (lukaszsus) [reply magda/module/module.py:102]: Ok :slightly_smiling_face:

File: test/utils/logger/printers/test_timestamp.py (hunk 15 above)
Comment (lukaszsus) [reply test/utils/logger/printers/test_timestamp.py]: Ok, thank you for your explanation. :-)

File: magda/pipeline/parallel/runtime.py (hunk 17 above)
Comment (jamebs) [reply magda/pipeline/parallel/runtime.py:61]: Only builder can call this method, so I marked it as semi-private

File: magda/pipeline/parallel/parallel_pipeline.py (hunk 3 above)
Comment (DanielPopek) [reply magda/pipeline/parallel/parallel_pipeline.py:40]: Okay, I understand your point. There is no bigger difference :)  You can lave it as `logger`

File: test/utils/logger/printers/test_group.py (hunk 13 above)
Comment (DanielPopek) [reply test/utils/logger/printers/test_group.py]: Quite a good idea 👍

**Review by DanielPopek — APPROVED** [review]

**Review by DanielPopek — DISMISSED** [review]

File: test/utils/logger/printers/test_group.py [hunk 18]
```diff
@@ -37,7 +37,7 @@ def test_should_accept_extra_arguments(self):
         replica = 20
 
         part = LoggerParts.Group(group, replica)
-        output = printer.flush(colors=False, group=part, extra=[1, 2, 3], msg='Test')
+        output_base = printer.flush(colors=False, group=part, msg='Test')
```
Comment (lukaszsus) [inline test/utils/logger/printers/test_group.py]: I think that `msg` is also an extra argument. I mean it is not printed too so I would remove it. It is up to you because finally, there is no difference for this test if you would use `output_base = printer.flush(colors=False, group=part)` instead of `output_base = printer.flush(colors=False, group=part, msg='Test')` but personally, I would rather change it to shorter version. The same in similar tests.

File: test/utils/logger/printers/test_message.py [hunk 19]
```diff
@@ -47,3 +54,12 @@ def test_event_should_be_different_than_standard_message(self):
         standard = printer.flush(colors=False, msg=base, is_event=False)
         event = printer.flush(colors=False, msg=base, is_event=True)
         assert standard != event
+
+    def test_should_accept_extra_arguments(self):
```
Comment (lukaszsus) [inline test/utils/logger/printers/test_message.py]: I know that I accepted this name in previous iteration of reviewing your pull request but I have noticed that maybe the name `test_should_ignore_extra_arguments` would describe this test better. What do you think? Of course, this decision also impacts on other tests with the same name.

File: test/utils/logger/printers/test_message.py [hunk 20]
```diff
@@ -34,6 +34,13 @@ def test_should_print_event(self):
         assert output.find(event) != -1
         assert re.search(self.COLOR_REGEXP, output) is None
 
+    def test_should_accept_empty_event(self):
+        printer = MessagePrinter()
+        event = ''
+        output = printer.flush(colors=False, msg=event, is_event=True)
+        assert output.find(event) != -1
```
Comment (lukaszsus) [inline test/utils/logger/printers/test_message.py]: I think that this assert makes no sense. Since `event=''` and `output` is a string, the expression `output.find(event)` will never be equal to `-1`. The same problem occurs in `test_should_accept_empty_message` with `assert output.find(message) != -1`.

**Review by lukaszsus — APPROVED** [review]
:ok_hand:

**Review by DanielPopek — APPROVED** [review]
